### PR TITLE
Introduce creation_time field into host description

### PR DESCRIPTION
**Pull request by nezdolik**
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Description:  Introducing "host creation time" field into host description. This change is prerequisite for supporting slow start in Envoy (https://github.com/envoyproxy/envoy/pull/13176) and as it touches lots of code, it was decided to factor it out into dedicated PR.
Risk Level: Medium
Testing: In progress
Docs Changes: NA

### CODE REVIEW

File: bazel/envoy_internal.bzl [hunk 1]
```diff
@@ -7,6 +7,9 @@ def envoy_copts(repository, test = False):
     posix_options = [
         "-Wall",
         "-Wextra",
+        "-Wno-sign-compare",
```
Comment (nezdolik) [inline bazel/envoy_internal.bzl:10]: this will be cleaned up

**Comment by mattklein123** [comment]
Let me know when this is ready for review!

/wait

**Comment by nezdolik** [comment]
Managed to track down duplicated time source issue in failing tests (original_dst, logical_dns tests) with gdb and fixed it. Updated all tests under `common/upstream` and some under `common/http`. Need to fix the rest of tests. But now is Friday and am signing off :)

**Comment by nezdolik** [comment]
my local machine clang formatter + buildifier formatted all bazel BUILD files (moved licence part at the top of file). Fixed that problem with running formatting in Docker

**Comment by nezdolik** [comment]
certain integration tests are timing out with non obvious error, will need to debug, could be because of multiple timesystems allocated in base integration test class.

**Comment by nezdolik** [comment]
Clang tidy is failing with memory leaks complaints, problems have been addressed here: https://github.com/envoyproxy/envoy/pull/14042. Once it gets merged will need to merge master here

**Comment by nezdolik** [comment]
@mattklein123 please take a look

**Comment by mattklein123** [comment]
Sorry can you merge main and then I will review?

**Comment by nezdolik** [comment]
@mattklein123 done

File: source/common/upstream/upstream_impl.h [hunk 2]
```diff
@@ -156,6 +158,7 @@ class HostDescriptionImpl : virtual public HostDescription,
   HealthCheckHostMonitorPtr health_checker_;
   std::atomic<uint32_t> priority_;
   Network::TransportSocketFactory& socket_factory_;
+  MonotonicTime creation_time_;
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.h:161]: nit: const

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 3]
```diff
@@ -55,9 +55,11 @@ RedisCluster::RedisCluster(
           factory_context.clusterManager(), factory_context.api().timeSource())),
       registration_handle_(refresh_manager_->registerCluster(
           cluster_name_, redirect_refresh_interval_, redirect_refresh_threshold_,
-          failure_refresh_threshold_, host_degraded_refresh_threshold_, [&]() {
+          failure_refresh_threshold_, host_degraded_refresh_threshold_,
+          [&]() {
             redis_discovery_session_.resolve_timer_->enableTimer(std::chrono::milliseconds(0));
-          })) {
+          })),
+      time_source_(dispatcher_.timeSource()) {
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc:62]: Can we have time_source_ in BaseDynamicClusterImpl or one of the other cluster base classes vs. having it in every derived class? That might be a bit simpler and cleaner?

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 4]
```diff
@@ -276,7 +278,8 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
     const int rand_idx = parent_.random_.random() % discovery_address_list_.size();
     auto it = discovery_address_list_.begin();
     std::next(it, rand_idx);
-    host = Upstream::HostSharedPtr{new RedisHost(parent_.info(), "", *it, parent_, true)};
+    host = Upstream::HostSharedPtr{
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc:281]: nit: while you are in here make_shared

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

File: source/extensions/clusters/redis/redis_cluster.cc (hunk 4 above)
Comment (nezdolik) [reply source/extensions/clusters/redis/redis_cluster.cc:281]: unfortunately cannot easily switch to make_shared here, as one of RedisHost base classes has std::atomic field, thus not copyable in make_shared unless custom copy constructor is provided.

**Comment by mattklein123** [comment]
Not sure what is up with CI but pretty sure this is not going to pass format.

/wait

**Comment by nezdolik** [comment]
> Not sure what is up with CI but pretty sure this is not going to pass format.
> 
> /wait

Oh, i had git rebase in progress and did not notice it, so one commit did not got pushed to remote...

**Comment by nezdolik** [comment]
build pipeline on my dev box started failing with gcc, trying to fix it now and rerun build and tests.

**Comment by nezdolik** [comment]
upgrading gcc,g++ to v 10 and purging bazel cache helped. I will keep pushing commits here to get faster feedback loop cycle. Am also running checks on devbox, but they are slower.

**Comment by nezdolik** [comment]
@mattklein123 PTAL

File: include/envoy/upstream/host_description.h [hunk 5]
```diff
@@ -148,6 +149,11 @@ class HostDescription {
    * Set the current priority.
    */
   virtual void priority(uint32_t) PURE;
+
+  /**
+   * @return timestamp in milliseconds of when host was created.
```
Comment (mattklein123) [inline include/envoy/upstream/host_description.h:154]: nit:

```suggestion
   * @return timestamp when host was created.
```

**Review by mattklein123 — APPROVED** [review]
Thanks! Feel free to fix the comment nit in your next change.

**Pull request merged**